[lldb] Move SBLanguages.h out of API tree

medismailben opened the pull request:
This patch moves `SBLanguages.h` out of the API tree. This file gets generated at build time using DWARF table-gen file and contains an enumeration of the DWARF supported source language names and there respective code/identifier.

Since this is an enum, this shouldn't be part of the API tree but rather it should be included in the `lldb-enumeration` header.

Also, that enum was used in internal methods in `lldb_private` by down-casting the enum value type to an `uint16_t`. This patch changes that by making those internal methods use the enum type.

This should address the feedbacks from #111907.

jimingham commented:
This looks better to me.  I'm not sure why we were using defines from llvm::DWARF that seem to overlap the ones here, however.  So I'll defer to Adrian or someone more knowledgeable about that side of things.

chelcassanova commented on lldb/scripts/generate-sbapi-dwarf-enum.py:
```diff
@@ -5,22 +5,21 @@
 
 namespace lldb {
-/// Used by \\ref SBExpressionOptions.
 /// These enumerations use the same language enumerations as the DWARF
```
This is a nit but maybe we want to add that this file is generated and that there's a reason this enum isn't directly implemented in `lldb-enumerations.h`?

medismailben commented:
> This looks better to me. I'm not sure why we were using defines from llvm::DWARF that seem to overlap the ones here, however. So I'll defer to Adrian or someone more knowledgeable about that side of things.

Yeah, that wasn't clear to me either, why didn't we just use the types in `llvm::dwarf::SourceLanguageName` and had to generate a new enum from the same table-gen file. @adrian-prantl any guesses ?

chelcassanova reviewed (commented):
This looks good to me and I agree with the original feedback that led to this change in the first place, though I'll also wait for any comments Adrian might have.

chelcassanova commented:
I'm just wondering, would this patch run into the same issue that I ran into on https://github.com/llvm/llvm-project/pull/111907 where this could cause a clean build to fail since the `SourceLangugageNames` file wouldn't be generated before it gets included in `lldb-enumerations.h`?

medismailben commented:
> I'm just wondering, would this patch run into the same issue that I ran into on #111907 where this could cause a clean build to fail since the `SourceLangugageNames` file wouldn't be generated before it gets included in `lldb-enumerations.h`?

Possibly, we'll see when it lands 😛

bulbazord commented on lldb/include/lldb/API/SBExpressionOptions.h:
```diff
@@ -71,7 +70,7 @@ class LLDB_API SBExpressionOptions {
   /// defined by the DWARF 6 specification.
   /// WARNING: These codes may change until DWARF 6 is finalized.
-  void SetLanguage(lldb::SBSourceLanguageName name, uint32_t version);
+  void SetLanguage(lldb::SourceLanguageName name, uint32_t version);
```
ABI break: Symbol changes names from `_ZN4lldb19SBExpressionOptions11SetLanguageENS_20SBSourceLanguageNameEj` to `_ZN4lldb19SBExpressionOptions11SetLanguageENS_18SourceLanguageNameEj`

jimingham replied:
I replied but not to this comment.  If the meanings of the enum might change, the API is explicitly not stable.

medismailben replied:
We can solve this by making making a typedef in SBDefine to keep the same symbol

bulbazord replied:
Changing the symbol name is an ABI break, changing the integral values would be an API break. My comment is only concerned about the former here.

labath replied:
Without weighing in on the patch itself, I want to say that this is not how I would understand this distinction. Although a change in the value of a constant would not result in symbol changes in lldb, the value can get embedded in the code of the lldb user, and even (e.g. through its use as a template arg) in the names of their symbols.

In terms of compatibility, I think there are two levels that we can talk about:
- whether existing source code will work when it is compiled against new version of lldb (compile time)
- whether existing compiled code will work when used with new lldb (at runtime)

Changing a constant value breaks the second use case (just like a symbol name change would), although it would be harder to detect (which, in a way, makes it worse).

bulbazord replied:
That's true. I'm fairly certain our "ABI stability" guarantee applies to both scenarios, though I'm not sure how we can promise an "unstable API" while maintaining our "ABI stability" guarantee. I want to review as many SBAPI changes as possible to look for these kinds of ABI breaks but it doesn't scale very well.

jimingham replied:
We don't have a formal policy about "provisional additions to the SB API".  The warning about the values change clearly means: "If you use this API, you will have to recompile against the version of lldb you intend to use."  So it is explicitly saying it's an unstable API.

Best would be to not do this, but if we find ourselves in a position where we have to, then we probably should require you to define `ALLOW_LLDB_UNSTABLE_SB_APIS` or some such, and guard unstable API's based on that.

jimingham commented:
I would feel a little worse about that if the API didn't explicitly say that the values of the enum might change meanings.  So it explicitly says it isn't stable...

medismailben committed:
[lldb] Move SBLanguages.h out of API tree

This patch moves `SBLanguages.h` out of the API tree. This file gets
generated at build time using DWARF table-gen file and contains an
enumeration of the DWARF supported source language names and there
respective code/identifier.

Since this is an enum, this shouldn't be part of the API tree but rather
it should be included in the `lldb-enumeration` header.

Also, that enum was used in internal methods in `lldb_private` by
down-casting the enum value type to an `uint16_t`.
This patch changes that by making those internal methods use the enum
type.

This should address the feedbacks from #111907.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

bulbazord commented on lldb/include/lldb/API/SBDefines.h:
```diff
@@ -136,6 +136,8 @@ class LLDB_API SBWatchpoint;
 class LLDB_API SBWatchpointOptions;
 class LLDB_API SBUnixSignals;
 
+typedef SourceLanguageName SBSourceLanguageName;
```
I'm not sure a typedef is enough here. It ends up being little more than syntactic sugar, the final symbol will still contain `SourceLanguageName` instead of `SBSourceLanguageName`. To verify this, you can perform a build with this patch and search for `SBExpressionOptions::SetLanguage` and make sure it still retains the `SBSourceLanguageName` symbol.
